### PR TITLE
DOC: move Array API tables to a user-visible location

### PR DESCRIPTION
**Pull request by ev-br**
#### Reference issue

closes https://github.com/scipy/scipy/issues/23398

#### What does this implement/fix?

Currently, the Array API tables are hidden in the developer documentation section. Move them in the Main SciPy namespace section, so that they are user-visible.

#### Additional information

As is, this PR anticipates making the support public, so does not mention the `SCIPY_ARRAY_API=1` environment variable. 
We'll also probably want to add a tutorial page, or at least a quick demo for the "array out = array in" principle. This is not done in this PR which just moves files around.

#### AI Generation Disclosure

No AI.

### CODE REVIEW

**Comment by lucascolley** [comment]
I think I would prefer adding this somewhere on https://scipy.github.io/devdocs/reference/index.html, this is still a bit hidden

**Comment by rgommers** [comment]
Thanks for working on this @ev-br! 

> I think I would prefer adding this somewhere on https://scipy.github.io/devdocs/reference/index.html, this is still a bit hidden

It's still a bit hard to find indeed (I had trouble locating it in the rendered docs for this PR, even though I knew it should be present). I personally don't have a strong opinion on whether it's a separate page, or in "main namespace", or in "index". The main thing it needs is discoverability, meaning it needs a prominent cross-link from the User Guide part of what's now hidden in the developer guide (_"Support for the array API standard"_). Moving that content is kinda linked to moving these tables, so let's make sure to do that in a follow-up PR right after this one goes in?

File: doc/source/reference/main_namespace.rst [hunk 1]
```diff
@@ -45,3 +45,150 @@ Submodules
 `special`      Special functions
 `stats`        Statistical functions
 =============  ==================================================
+
+
+
```
Comment (DietBru) [inline doc/source/reference/main_namespace.rst:50]: I was not aware of this PR when I was restructuring the "API reference" front page and deleting the `main_namespace.rst` file (it was kind of hidden away anyways...), sorry.

The way to move forward could be to move the following sections to end of the file `reference/index.rst` (after rebasing onto current main).

**Comment by ev-br** [comment]
Rebased on main and moved the tables to the main refernce page, https://output.circle-artifacts.com/output/job/7f6ebaaf-38fa-47ff-ae75-c2734864e390/artifacts/0/html/reference/index.html#scipy-api

Am ready to move it to a separate page after `index`, if a general feeling is that it should be separate.

Quick notes:

- "Array API support" sections in individual docstrings link to the new section of the main reference docs; 
- The `scipy.signal` implementation caveats stay in the development area, https://output.circle-artifacts.com/output/job/7f6ebaaf-38fa-47ff-ae75-c2734864e390/artifacts/0/html/dev/api-dev/array_api.html ; that page needs a rework anyway, kept is out of scope;
- did not systematically remove mentions of the `SCIPY_ARRAY_API` environment variable. This is best done when we actually remove it.

**Comment by j-bowhay** [comment]
Part of me wonders if we put it on its own page with its own link in the navbar to really shout about it. Normal when I click on the "SciPy API" page I then go straight to "Section Navigation" so I wouldn't have scrolled down far enough to see this exists.

**Comment by mdhaber** [comment]
I can see this going in the top navbar, too. To make space, maybe replace "building from source"? I'd think that could be covered by "Installing" and "Development".
Maybe something mentioning the libraries (e.g. "For CuPy/PyTorch/JAX Users" or "Use w/ CuPy/PyTorch/JAX") would be more recognizable than something mentioning "Python array API".

If it stays at the API reference, please consider removing the entries from the left panel:
image

File: doc/source/dev/api-dev/array_api.rst [hunk 2]
```diff
@@ -1401,128 +1401,38 @@ helped during the development phase:
...
+Hence, there can be, especially during CI testing, discrepancies in behavior between
+the default NumPy-based implementation and the JAX and CuPy backends. Skipping the
+incompatible backends in unit tests, as described in the
+:ref:`dev-arrayapi_adding_tests` section, is the currently recommended workaround.
+
+The functions are decorated by the code in file
+``scipy/signal/_support_alternative_backends.py``:
+
+.. literalinclude:: ../../../../scipy/signal/_support_alternative_backends.py
+    :lineno-match:
```
Comment (mdhaber) [inline doc/source/dev/api-dev/array_api.rst:1428]: `literalinclude` adds quite a bit to the page. I suppose it doesn't matter,  since it's at the bottom, but I think a link to the file or an illustrative excerpt would do the trick. (I see that this was just copy-pasted over, but maybe worth reconsidering now.)

File: doc/source/reference/index.rst [hunk 3]
```diff
@@ -142,3 +142,161 @@ Main namespace (``scipy``)
...
+.. code:: python
+
+    >>> import torch
+    >>> from scipy.cluster.vq import vq
+    >>> code_book = torch.tensor([[1., 1., 1.],
+    ...                           [2., 2., 2.]])
+    >>> features  = torch.tensor([[1.9, 2.3, 1.7],
+    ...                           [1.5, 2.5, 2.2],
+    ...                           [0.8, 0.6, 1.7]])
+    >>> code, dist = vq(features, code_book)
```
Comment (mdhaber) [inline doc/source/reference/index.rst:172]: May be out of scope, but there may be better examples of "array API compatibility", since `vq` doesn't rely much on the array API - it converts to NumPy, runs the code, and converts back. 
If desired, I could try to find an example that sells alternative backends a bit better. For instance, `logsumexp` can give a speedup with PyTorch, JAX (with JIT), and CuPy; it also supports forward-mode autodiff with JAX and PyTorch that can be checked against `softmax`.

**Comment by lucascolley** [comment]
> I can see this going in the top navbar, too. To make space, maybe replace "building from source"? I'd think that could be covered by "Installing" and "Development".

Maybe, but I'd say not quite yet at least. https://github.com/scipy/scipy/pull/26015 is the next open PR in a stack which has been separating user-/distributor-facing building docs from contributor-facing building docs. There is further work to be done on updating the docs on system-level dependencies after / as part of that.

I'm not sure what I think yet, but I can see the argument for keeping the building nav, if only for the section targeting distributors: https://output.circle-artifacts.com/output/job/f794f534-fc83-4236-8abb-5e7f4c25e2b5/artifacts/0/html/building/index.html#customizing-builds.

An alternative option would be to remove 'Installing' from the nav, and just have it linked prominently at the top of both the homepage and the user guide -- that sounds sufficient to me?

File: doc/source/reference/index.rst (hunk 3 above)
Comment (ev-br) [reply doc/source/reference/index.rst:172]: Agreed, a better demonstration would be nice here. This one I simply copied from the existing page, https://docs.scipy.org/doc/scipy/dev/api-dev/array_api.html

File: doc/source/dev/api-dev/array_api.rst (hunk 2 above)
Comment (ev-br) [reply doc/source/dev/api-dev/array_api.rst:1428]: This part is taken verbatim from the existing page, 
https://docs.scipy.org/doc/scipy/dev/api-dev/array_api_modules_tables/signal.html#array-api-support-signal-cpu
On main, it's in the `scipy.signal` specific dev page, now that this PR moves submodule-specific pages to the user-facing API reference area, I simply copy-pasted it to the development page. 
I should probably make the authorship of the commit back to @DietBru who wrote the original page.
